src/main/java/net/snowflake/client/core/JsonSqlInput.java

```diff
@@ -187,47 +209,118 @@ public Timestamp readTimestamp(TimeZone tz) throws SQLException {
           if (value == null) {
             return null;
           }
-          int columnType = ColumnTypeHelper.getColumnType(fieldMetadata.getType(), session);
-          int columnSubType = fieldMetadata.getType();
-          int scale = fieldMetadata.getScale();
-          Timestamp result =
-              SqlInputTimestampUtil.getTimestampFromType(
-                  columnSubType, (String) value, session, sessionTimeZone, tz);
-          if (result != null) {
-            return result;
+          return formatTimestamp(tz, value, fieldMetadata);
+        });
+  }
+
+  @Override
+  public <T> T readObject(Class<T> type, TimeZone tz) throws SQLException {
+    return withNextValue(
+        (value, jsonNode, fieldMetadata) -> {
+          if (SQLData.class.isAssignableFrom(type)) {
+            if (jsonNode.isNull()) {
+              return null;
+            } else {
+              SQLInput sqlInput =
+                  new JsonSqlInput(
+                      jsonNode, session, converters, fieldMetadata.getFields(), sessionTimeZone);
+              SQLData instance = (SQLData) SQLDataCreationHelper.create(type);
+              instance.readSQL(sqlInput, null);
+              return (T) instance;
+            }
+          } else if (Map.class.isAssignableFrom(type)) {
+            if (value == null) {
+              return null;
+            } else {
+              return (T) convertSqlInputToMap((SQLInput) value);
+            }
+          } else if (String.class.isAssignableFrom(type)
+              || Boolean.class.isAssignableFrom(type)
+              || Byte.class.isAssignableFrom(type)
+              || Short.class.isAssignableFrom(type)
+              || Integer.class.isAssignableFrom(type)
+              || Long.class.isAssignableFrom(type)
+              || Float.class.isAssignableFrom(type)
+              || Double.class.isAssignableFrom(type)) {
+            if (value == null) {
+              return null;
+            }
+            return (T) value;
+          } else if (Date.class.isAssignableFrom(type)) {
+            if (value == null) {
```

Comment:
Maybe we can move this condition to the very beginning of the function?

Reply:
It isn't verified in first of if blocks.

Reply:
So maybe we can have it like this:

```
if (SQLData.class.isAssignableFrom(type)) {
... 
} else {
   if (value == null) {
       return null;
   } else if (Map.class.isAssignableFrom(type)) {
   ......

}

Reply:
Fixed